core/kernel.go

```diff
@@ -72,292 +62,197 @@ const (
+}
 
-	logger = logger.WithScope("NewKernel()").With(structure.TimeKey, log.DefaultTimestampUTC,
-		structure.RunId, kern.RunID.String())
-	tmLogger := logger.With(structure.CallerKey, log.Caller(LoggingCallerDepth+1))
+// SetValidator sets the identity for the kernel
+func (kern *Kernel) SetValidator(privValidator tmTypes.PrivValidator) error {
+	var err error
+	kern.RunID, err = simpleuuid.NewTime(time.Now()) // Create a random ID based on start time
+	if err != nil {
+		return err
+	}
+	kern.View = tendermint.NewNodeView(kern.txCodec, kern.RunID)
+	err = kern.View.SetValidator(privValidator)
+	return err
+}
+
+// SetLogger initializes the kernel with the provided logger
+func (kern *Kernel) SetLogger(logger *logging.Logger) {
+	logger = logger.WithScope("NewKernel()").With(structure.TimeKey,
+		log.DefaultTimestampUTC, structure.RunId, kern.RunID.String())
 	kern.Logger = logger.WithInfo(structure.CallerKey, log.Caller(LoggingCallerDepth))
-	stateDB := NewBurrowDB(tmConf.DBDir())
+	kern.Emitter = event.NewEmitter(kern.Logger)
+}
 
+// LoadState starts from scratch or previous chain
+func (kern *Kernel) LoadState(genesisDoc *genesis.GenesisDoc, dbDir, restore string) (err error) {
+	stateDB := newBurrowDB(dbDir)
 	kern.Blockchain, err = bcm.LoadOrNewBlockchain(stateDB, genesisDoc, kern.Logger)
 	if err != nil {
-		return nil, fmt.Errorf("error creating or loading blockchain state: %v", err)
+		return fmt.Errorf("error creating or loading blockchain state: %v", err)
 	}
+	kern.Blockchain.SetBlockStore(bcm.NewBlockStore(blockchain.NewBlockStore(stateDB)))
 
 	// These should be in sync unless we are at the genesis block
 	if kern.Blockchain.LastBlockHeight() > 0 {
 		kern.Logger.InfoMsg("Loading application state")
 		kern.State, err = state.LoadState(stateDB, execution.VersionAtHeight(kern.Blockchain.LastBlockHeight()))
 		if err != nil {
-			return nil, fmt.Errorf("could not load persisted execution state at hash 0x%X: %v",
+			return fmt.Errorf("could not load persisted execution state at hash 0x%X: %v",
 				kern.Blockchain.AppHashAfterLastBlock(), err)
 		}
 		if !bytes.Equal(kern.State.Hash(), kern.Blockchain.AppHashAfterLastBlock()) {
-			return nil, fmt.Errorf("state and blockchain disagree on hash for block at height %d: "+
+			return fmt.Errorf("state and blockchain disagree on hash for block at height %d: "+
 				"state gives %X, blockchain gives %X", kern.Blockchain.LastBlockHeight(),
 				kern.State.Hash(), kern.Blockchain.AppHashAfterLastBlock())
 		}
 		if restore != "" {
-			return nil, fmt.Errorf("Cannot restore onto existing chain; don't give --restore-dump argument")
+			return fmt.Errorf("Cannot restore onto existing chain; don't give --restore-dump argument")
 		}
 	} else {
 		kern.State, err = state.MakeGenesisState(stateDB, genesisDoc)
 		if err != nil {
-			return nil, fmt.Errorf("could not build genesis state: %v", err)
+			return fmt.Errorf("could not build genesis state: %v", err)
 		}
 
 		if restore != "" {
```

Comment:
Let's remove the restore flag and make `RestoreDump` it's own function on `Kernel`

Reply:
Done :)